### PR TITLE
HBASE-23234 Provide .editorconfig based on checkstyle configuration

### PR DESCRIPTION
**Pull request by ndimiduk**
This file is generated using IntelliJ, following these steps:
 #. Open Preferences > Editor > Code Style
 #. Select (config wheel) > Import Schema > CheckStyle Configuration
 #. Select `hbase-checkstyle/src/main/resources/hbase/checkstyle.xml`
 #. Select (config wheel) > Export > EditorConfig File

### CODE REVIEW

**Comment by ndimiduk** [comment]
Here's an experiment with [EditorConfig](https://editorconfig.org), a supposedly tool-neutral code formatting specification. I've generated this configuration from our existing CheckStyle configuration, following the steps outlined in the commit. As reviewers, I've added people who have touched the existing `dev-support/hbase_eclipse_formatter.xml` definition. I'll try it out in IntelliJ over the coming days/weeks and see how it goes. Any Eclipse users out there mind giving it a go as well?

**Comment by busbey** [comment]
how do we keep this up to date if checkstyle changes? any way we can check it via cli? then we could at least flag up that a sync was needed?

**Comment by busbey** [comment]
(question is not meant to be a suggest we don't move forward with this, fyi. we have no way of checking the existing tool specific formatters are up to date)

**Comment by bharathv** [comment]
One question is what if the checkstyle file changes, how do we keep both the files in sync?

**Comment by ndimiduk** [comment]
Yeah I agree totally.

Maybe the first question is: should checkstyle be our source of truth on this stuff? I'm guessing the answer is yes.

Maybe we can add a bit to our Yetus personality that looks for changes to either file to include changes in both?

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 29s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 16s |  The patch does not generate ASF License warnings.  |
|  |   |   1m 54s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.4 Server=19.03.4 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-846/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/846 |
| Optional Tests | dupname asflicense |
| uname | Linux 2aa3b279655f 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-846/out/precommit/personality/provided.sh |
| git revision | master / 8bfdfe1b85 |
| Max. process+thread count | 41 (vs. ulimit of 10000) |
| modules | C: . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-846/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by ndimiduk** [comment]
I see all these `ij_` prefixed lines are specific to IntelliJ 😞

**Comment by bharathv** [comment]
Sean beat me to it. Looks like he put the same comment. I'm an intellij user but just out of curiosity I tried eclipse to see how the whole checkstyle thing works. Following are my observations fwiw.

1. I tried eclipse-cs [1]. It fails with the following error which apparently is a known problem https://github.com/checkstyle/eclipse-cs/issues/107. Couldn't workaround it.

java.lang.NoClassDefFoundError: org/eclipse/jdt/internal/ui/preferences/PreferencesAccess
        at net.sf.eclipsecs.core.transformer.FormatterConfigWriter.writeCleanupSettings(FormatterConfigWriter.java:95)
        at net.sf.eclipsecs.core.transformer.FormatterConfigWriter.writeSettings(FormatterConfigWriter.java:89)
        at net.sf.eclipsecs.core.transformer.FormatterConfigWriter.<init>(FormatterConfigWriter.java:81)
        at net.sf.eclipsecs.core.transformer.CheckstyleTransformer.transformRules(CheckstyleTransformer.java:124)
        at net.sf.eclipsecs.core.jobs.TransformCheckstyleRulesJob.runInWorkspace(TransformCheckstyleRulesJob.java:119)
        at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:42)
        at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)

2. This patch's editorconfig file doesn't work with my eclipse setup for whatever reason. I'm not able to figure out why.

[1] https://stackoverflow.com/questions/984778/how-to-generate-an-eclipse-formatter-configuration-from-a-checkstyle-configurati

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,693 @@
...
+ij_java_field_annotation_wrap = normal
+ij_java_finally_on_new_line = false
+ij_java_for_brace_force = always
+ij_java_for_statement_new_line_after_left_paren = false
+ij_java_for_statement_right_paren_on_new_line = false
+ij_java_for_statement_wrap = off
+ij_java_generate_final_locals = false
+ij_java_generate_final_parameters = false
+ij_java_if_brace_force = always
+ij_java_imports_layout = $*,*,org.apache.hbase.thirdparty.**,org.apache.hadoop.hbase.shaded.**
```
Comment (busbey) [inline .editorconfig]: nice!

Comment (busbey) [reply .editorconfig]: FWIW, the import order seems to be the thing that trips up IDE users the most. so anything we can do to make that better for folks is A+ by me.

Comment (ndimiduk) [reply .editorconfig]: I agree on this point. Now, is this import order actually correct..? What about import whitespace?

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
||| _ master Compile Tests _ |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 16s |  The patch does not generate ASF License warnings.  |
|  |   |   1m 58s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.4 Server=19.03.4 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-846/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/846 |
| Optional Tests | dupname asflicense |
| uname | Linux 7b69215259cc 4.15.0-66-generic #75-Ubuntu SMP Tue Oct 1 05:24:09 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-846/out/precommit/personality/provided.sh |
| git revision | master / 4ea792246f |
| Max. process+thread count | 42 (vs. ulimit of 10000) |
| modules | C: . U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-846/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**